rustls/src/client/hs.rs

```diff
@@ -1181,6 +1189,33 @@ impl State<ClientConnectionData> for ExpectServerHelloOrHelloRetryRequest {
     }
 }
 
+#[derive(Debug)]
+struct CertificateTypeExtensionParams {
+    server_negotiated: Option<CertificateType>,
+    client_expects: CertificateType,
+    extension_type: ExtensionType,
+}
+
+impl CertificateTypeExtensionParams {
+    fn validate_certificate_type_extension(&self) -> CertificateTypeExtensionNegotationResult {
+        let server_supports_raw_key = self.server_negotiated == Some(CertificateType::RawPublicKey);
+        let client_expects_raw_key = self.client_expects == CertificateType::RawPublicKey;
+        match (client_expects_raw_key, server_supports_raw_key) {
+            (true, true) => CertificateTypeExtensionNegotationResult::Negotiated(
+                self.extension_type,
+                CertificateType::RawPublicKey,
+            ),
+            (false, false) => CertificateTypeExtensionNegotationResult::NotNegotiated,
+            (true, false) => CertificateTypeExtensionNegotationResult::Err(
+                Error::PeerIncompatible(PeerIncompatible::IncorrectCertificateTypeExtension),
+            ),
+            (false, true) => CertificateTypeExtensionNegotationResult::Err(
```

Comment:
The coverage gap flagged for this branch seems worth considering if it isn't too tricky. WDYT?

Reply:
To be honest I'm not sure if it's possible to get to the `(false, true)` arm. 
There are already tests sending an unsolicited certificate type extension in the `ServerHello`, but then an `UnsolicitedEncryptedExtension` error is raised. 
If there is truly no way to get to that situation we could also remove the `UnsolicitedCertificateTypeExtension` of course.